pkg/git/service.go

```diff
@@ -188,6 +190,65 @@ func (s *Service) SetGitConfig(userData *gitprovider.GitUser) error {
 	return nil
 }
 
+func (s *Service) GetUnpushedCommitsCount(branchName string) (int, error) {
+	r, err := git.PlainOpen(s.ProjectDir)
+	if err != nil {
+		return 0, err
+	}
+
+	mainRef, err := r.Reference(plumbing.ReferenceName(fmt.Sprintf("refs/heads/%s", branchName)), true)
+	if err != nil {
+		return 0, err
+	}
+	originMainRef, err := r.Reference(plumbing.ReferenceName(fmt.Sprintf("refs/remotes/origin/%s", branchName)), true)
+	if err != nil {
+		return 0, err
+	}
+
+	mainCommits, err := s.getCommitsInBranch(r, mainRef.Hash())
+	if err != nil {
+		return 0, err
+	}
+
+	originMainCommits, err := s.getCommitsInBranch(r, originMainRef.Hash())
+	if err != nil {
+		return 0, err
+	}
+
+	originMainCommitMap := make(map[plumbing.Hash]bool)
```

Comment:
While this iteration over all commits might work in theory, in practice, where branches can have thousands of commits, it's not really viable.

I suggest that you change the approach. For example, you start iterating on he commits on the `mainRef` and stop when you reach the same commit on the `originMainCommits` branch. This is just a thought.

Maybe you can look through how some Go git libraries implement this for inspiration.